### PR TITLE
JBPM-9992 Extended LogCleanupCommand to clean up ExecutionErrorInfo

### PR DESCRIPTION
**Pull request by hbelmiro**
JIRA: https://issues.redhat.com/browse/JBPM-9992
Depends on: https://github.com/kiegroup/droolsjbpm-knowledge/pull/588

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

File: jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/commands/ExecutionErrorCleanupCommand.java [hunk 1]
```diff
@@ -16,6 +16,7 @@
 
 package org.jbpm.executor.commands;
 
+import java.text.ParseException;
```
Comment (gmunozfe) [inline jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/commands/ExecutionErrorCleanupCommand.java:19]: not used, remove it

File: jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/commands/LogCleanupCommand.java [hunk 2]
```diff
@@ -16,6 +16,7 @@
 
 package org.jbpm.executor.commands;
 
+import java.text.ParseException;
```
Comment (gmunozfe) [inline jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/commands/LogCleanupCommand.java:19]: Not used, remove it

File: jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/commands/ExecutionErrorCleanup.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        Map<String, Object> parameters = new HashMap<>();
+        StringBuilder cleanUpErrorsQuery = new StringBuilder();
+
+        cleanUpErrorsQuery.append("delete from ExecutionErrorInfo where (processInstanceId in ")
+                          .append("(select processInstanceId from ProcessInstanceLog where status in (")
+                          .append(ProcessInstance.STATE_COMPLETED)
+                          .append(",")
+                          .append(ProcessInstance.STATE_ABORTED)
+                          .append("))")
+                          .append(" or processInstanceId not in (select processInstanceId from ProcessInstanceLog))");
```
Comment (gmunozfe) [inline jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/commands/ExecutionErrorCleanup.java:73]: This is to remove orphan processInstanceIds, not existing at ProcessInstanceLog, isn't it? is this scenario reachable?

**Review by gmunozfe — APPROVED** [review]
Looks good to me, great work @hbelmiro !
Just some minor comments and one question in the modified query.

File: jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/commands/ExecutionErrorCleanup.java (hunk 3 above)
Comment (hbelmiro) [reply jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/commands/ExecutionErrorCleanup.java:73]: Yes, it is. I reached that scenario by the new test `LogCleanupCommandTest#notSkipExecutorLog`. I'm not sure if this is reachable by production code though.

**Comment by hbelmiro** [comment]
Many thanks for the review @gmunozfe.
I committed the suggested changes.

File: jbpm-services/jbpm-executor-cdi/src/test/filtered-resources/META-INF/persistence.xml [hunk 4]
```diff
@@ -10,6 +10,8 @@
     <mapping-file>META-INF/Executor-orm.xml</mapping-file>
     <class>org.jbpm.executor.entities.ErrorInfo</class>
     <class>org.jbpm.executor.entities.RequestInfo</class>
+    <class>org.jbpm.process.audit.ProcessInstanceLog</class>
+    <class>org.jbpm.runtime.manager.impl.jpa.ExecutionErrorInfo</class>
```
Comment (elguardian) [inline jbpm-services/jbpm-executor-cdi/src/test/filtered-resources/META-INF/persistence.xml:14]: why is this here if the test is in test coverage ?

File: jbpm-services/jbpm-executor/src/test/filtered-resources/META-INF/persistence.xml [hunk 5]
```diff
@@ -10,6 +10,8 @@
     <mapping-file>META-INF/Executor-orm.xml</mapping-file>
     <class>org.jbpm.executor.entities.ErrorInfo</class>
     <class>org.jbpm.executor.entities.RequestInfo</class>
+    <class>org.jbpm.process.audit.ProcessInstanceLog</class>
```
Comment (elguardian) [inline jbpm-services/jbpm-executor/src/test/filtered-resources/META-INF/persistence.xml:13]: why is this here if the test is in test coverage ?

File: jbpm-services/jbpm-executor-cdi/src/test/filtered-resources/META-INF/persistence.xml (hunk 4 above)
Comment (hbelmiro) [reply jbpm-services/jbpm-executor-cdi/src/test/filtered-resources/META-INF/persistence.xml:14]: The entities need to be mapped, otherwise they can't be found when we run the query.
I didn't get what you mean with:
> ... if the test is in test coverage

File: jbpm-services/jbpm-executor/src/test/filtered-resources/META-INF/persistence.xml (hunk 5 above)
Comment (hbelmiro) [reply jbpm-services/jbpm-executor/src/test/filtered-resources/META-INF/persistence.xml:13]: The entities need to be mapped, otherwise they can't be found when we run the query.
I didn't get what you mean with:
> ... if the test is in test coverage

**Comment by hbelmiro** [comment]
Jenkins retest this

File: jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/commands/ExecutionErrorCleanupCommand.java [hunk 6]
```diff
@@ -99,44 +86,7 @@ public ExecutionResults execute(CommandContext ctx) throws Exception {
...
-        }
-        if (forDeployment != null && !forDeployment.isEmpty()) {            
-            cleanUpErrorsQuery.append(" and deploymentId = :forDeployment");
-            parameters.put("forDeployment", forDeployment);
-        }
-        
-        TransactionalCommandService commandService = new TransactionalCommandService(emf);
-        
-        int deletedErrors = commandService.execute(new UpdateStringCommand(cleanUpErrorsQuery.toString(), parameters));
+		int deletedErrors = new ExecutionErrorCleanup(emf).cleanup(ctx);
```
Comment (fjtirado) [inline jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/commands/ExecutionErrorCleanupCommand.java:89]: Can we change this to static usage, I mean, use ExecutionErrorCleanupHelper.cleanup(emf, ctx)

**Comment by hbelmiro** [comment]
jenkins retest this

File: jbpm-services/jbpm-executor-cdi/src/test/filtered-resources/META-INF/persistence.xml (hunk 4 above)
Comment (elguardian) [reply jbpm-services/jbpm-executor-cdi/src/test/filtered-resources/META-INF/persistence.xml:14]: Hi, 
This should be consider a smell because you are mapping an entity that does not belong to this package/module here. This should tip you that you are breaking module coherence.
I understand because you are modifying the persistence here it needs process  instance log (not defined here) and Execution error info (not defined here either). Same comment as below.

Comment (hbelmiro) [reply jbpm-services/jbpm-executor-cdi/src/test/filtered-resources/META-INF/persistence.xml:14]: Got your point now. Nice catch.
Thanks.

Comment (hbelmiro) [reply jbpm-services/jbpm-executor-cdi/src/test/filtered-resources/META-INF/persistence.xml:14]: @elguardian wouldn't the above mappings also be a smell? They're from another package as well.

````xml
    <class>org.jbpm.executor.entities.ErrorInfo</class>
    <class>org.jbpm.executor.entities.RequestInfo</class>
````

**Comment by hbelmiro** [comment]
@elguardian I reverted the original command (the one that would have its compatibility broken) and created a service as suggested like https://github.com/kiegroup/jbpm/blob/5348ff384ac841593d487790f24dc56a64f17478/jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/commands/LogCleanupCommand.java#L153-L161
Now we have the entity not mapped error.

````
java.lang.IllegalArgumentException: org.hibernate.hql.internal.ast.QuerySyntaxException: ExecutionErrorInfo is not mapped [DELETE FROM ExecutionErrorInfo l]
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:138)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:729)
	at org.hibernate.internal.AbstractSessionImpl.createQuery(AbstractSessionImpl.java:23)
	at org.jbpm.process.audit.JPAAuditLogService.executeQuery(JPAAuditLogService.java:446)
	at org.jbpm.process.audit.JPAAuditLogService.doDelete(JPAAuditLogService.java:408)
	at org.jbpm.process.audit.query.AbstractAuditDeleteBuilderImpl$2.execute(AbstractAuditDeleteBuilderImpl.java:254)
	at org.jbpm.executor.commands.LogCleanupCommand.execute(LogCleanupCommand.java:180)
	at org.jbpm.executor.impl.AbstractAvailableJobsExecutor.executeGivenJob(AbstractAvailableJobsExecutor.java:136)
	at org.jbpm.executor.impl.AvailableJobsExecutor.executeJob(AvailableJobsExecutor.java:33)
	at org.jbpm.executor.impl.concurrent.PrioritisedRunnable.run(PrioritisedRunnable.java:47)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at org.jbpm.executor.impl.concurrent.PrioritisedScheduledFutureTask.run(PrioritisedScheduledFutureTask.java:41)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
````

**Comment by elguardian** [comment]
strange... the dependency     
   <dependency>
      <groupId>org.jbpm</groupId>
      <artifactId>jbpm-runtime-manager</artifactId>
      <optional>true</optional>
    </dependency>
is in the pom (but as optional)

it is also mapped in during test.
https://github.com/kiegroup/jbpm/blob/main/jbpm-services/jbpm-executor/src/test/filtered-resources/META-INF/persistence.xml#L81

**Comment by elguardian** [comment]
btw... I am executing LogCleanupCommandTest and it is not working
java.lang.IllegalArgumentException: org.hibernate.QueryException: could not resolve property: date of: org.jbpm.runtime.manager.impl.jpa.ExecutionErrorInfo [DELETE FROM org.jbpm.runtime.manager.impl.jpa.ExecutionErrorInfo l WHERE ( l.date <= :A1 )]

Could you point out where are you finding that problem ?

**Comment by hbelmiro** [comment]
@elguardian the test that is failing with entity not mapped error is `org.jbpm.executor.NoCDISimpleExecutorTest#cleanupLogExecutionTest`. This test uses the PersistenceUnit "org.jbpm.executor" which is defined in https://github.com/hbelmiro/jbpm/blob/64bc606f17ba0e100058900e6ac5c1ccb13486e5/jbpm-services/jbpm-executor/src/test/filtered-resources/META-INF/persistence.xml#L7

> btw... I am executing LogCleanupCommandTest and it is not working java.lang.IllegalArgumentException: org.hibernate.QueryException: could not resolve property: date of: org.jbpm.runtime.manager.impl.jpa.ExecutionErrorInfo [DELETE FROM org.jbpm.runtime.manager.impl.jpa.ExecutionErrorInfo l WHERE ( l.date <= :A1 )]

I hadn't notice that error. I did a [new commit](https://github.com/kiegroup/jbpm/pull/2086/commits/a67548b6d996f05393e5273f5a90de8dadaaf266) and this error shouldn't occur anymore.

**Comment by gmunozfe** [comment]
jenkins retest this
